### DISCUSSION TITLE
Support Multi channel audio for Opus for mpegts (SRT) stream

### DISCUSSION THREAD

**Discussion by ditogam** [bluenviron/mediamtx#3353] [General]
Hi everyone and thank you very much for such amazing produc you done.

I have little feature request. If we stream multi channel Opus audio (5.1 or 7.1) sound is damaged on recievers side, of course multi channel is not supported with webrtc, but rtsp and srt can handle it.

After only few lines changing in gortsplib and mediamtx, it working. It would be nice to have this feature in main distribution as well.

Thank you

```

Subject: [PATCH] gortsplib
---
Index: pkg/format/opus.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/format/opus.go b/pkg/format/opus.go
--- a/pkg/format/opus.go	(revision 5eddb779359c48c8f148f56e37cfce048c1078f3)
+++ b/pkg/format/opus.go	(date 1715121821736)
@@ -15,6 +15,7 @@
 type Opus struct {
 	PayloadTyp uint8
 	IsStereo   bool
+	Channels   int
 }
 
 func (f *Opus) unmarshal(ctx *unmarshalContext) error {
```
and mediamtx 
```
Subject: [PATCH] mediamtx
---
Index: internal/protocols/mpegts/to_stream.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/internal/protocols/mpegts/to_stream.go b/internal/protocols/mpegts/to_stream.go
--- a/internal/protocols/mpegts/to_stream.go	(revision dcb5b45e84c9667f3b42c98cd09fca9ccfe54fc3)
+++ b/internal/protocols/mpegts/to_stream.go	(date 1715216495245)
@@ -114,6 +114,7 @@
 				Formats: []format.Format{&format.Opus{
 					PayloadTyp: 96,
 					IsStereo:   (codec.ChannelCount >= 2),
+					Channels:   codec.ChannelCount,
 				}},
 			}
 
Index: internal/protocols/mpegts/from_stream.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/internal/protocols/mpegts/from_stream.go b/internal/protocols/mpegts/from_stream.go
--- a/internal/protocols/mpegts/from_stream.go	(revision dcb5b45e84c9667f3b42c98cd09fca9ccfe54fc3)
+++ b/internal/protocols/mpegts/from_stream.go	(date 1715216495270)
@@ -166,10 +166,7 @@
 			case *format.Opus:
 				track := addTrack(&mcmpegts.CodecOpus{
 					ChannelCount: func() int {
-						if forma.IsStereo {
-							return 2
-						}
-						return 1
+						return forma.Channels
 					}(),
 				})
 

```

**Comment by abalabolov** [comment]
We are also looking forward to this update.

**Comment by aler9** [comment]
Hello @ditogam, thanks for reporting the issue and posting the patch.

The reason why there's no `Channel` field inside `formats.Opus` is that formats are RTSP-specific entities, and in the Opus format there's no place in which channel count can be set or received, according to the specification:

https://datatracker.ietf.org/doc/html/rfc7587

In the specification you can see that there's a place for setting a "stereo" flag (`sprop-stereo`) but not a place for setting a "channel count". Therefore, if a multi channel audio track is streamed through RTSP, the channel count is lost or wrongly set to either 1 or 2.

Your patch covers the MPEGTS-to-MPEGTS case, but not the RTSP-to-MPEGTS case.

Therefore, i don't think that hacking `formats.Opus` is a valid solution, but we can think about other means for extracting and setting the channel count in protocols that require it to be explicitly provided (like MPEG-TS), for instance, by analyzing the content of the stream. I've created a dedicated issue here: #3355

**Reply by ditogam** [reply]
@aler9 Thank you for answer and for creating issue

[a comment by an automated account is not shown]
